File: rider_app/lib/core/api/models/api_response.dart
import '../api_helpers.dart';

/// Standard API Response wrapper (`ApiResponse` / `ApiResponse<T>`).
class ApiResponse {
  final int? status;
  final bool success;
  final String? message;
  final String? errorDetail;
  final String? code;

  const ApiResponse({
    this.status,
    required this.success,
    this.message,
    this.errorDetail,
    this.code,
  });

  factory ApiResponse.fromJson(Map<String, dynamic> json) {
    return ApiResponse(
      status: readField<num>(json, 'Status')?.toInt(),
      success: readField<bool>(json, 'Success') ?? false,
      message: readField<String>(json, 'Message'),
      errorDetail: readField<String>(json, 'ErrorDetail'),
      code: readField<String>(json, 'Code'),
    );
  }
}

class ApiResponseValue<T> {
  final int? status;
  final bool success;
  final String? message;
  final String? errorDetail;
  final String? code;
  final T? value;

  const ApiResponseValue({
    this.status,
    required this.success,
    this.message,
    this.errorDetail,
    this.code,
    this.value,
  });

  factory ApiResponseValue.fromJson(
    Map<String, dynamic> json,
    T Function(Map<String, dynamic>) fromJsonT,
  ) {
    final rawValue = readField<dynamic>(json, 'Value') ?? readField<dynamic>(json, 'value');
    return ApiResponseValue(
      status: readField<num>(json, 'Status')?.toInt(),
      success: readField<bool>(json, 'Success') ?? false,
      message: readField<String>(json, 'Message'),
      errorDetail: readField<String>(json, 'ErrorDetail'),
      code: readField<String>(json, 'Code'),
      value: rawValue == null ? null : fromJsonT(asMap(rawValue)),
    );
  }
}

class PaginatedResult<T> {
  final List<T> items;
  final int totalCount;
  final int page;
  final int pageSize;
  final bool hasPrevious;
  final bool hasNext;

  const PaginatedResult({
    required this.items,
    required this.totalCount,
    required this.page,
    required this.pageSize,
    required this.hasPrevious,
    required this.hasNext,
  });

  factory PaginatedResult.fromJson(
    Map<String, dynamic> json,
    T Function(Map<String, dynamic>) fromJsonT,
  ) {
    final rawItems = readField<List>(json, 'Items') ?? readField<List>(json, 'items') ?? [];
    return PaginatedResult(
      items: rawItems.map((e) => fromJsonT(asMap(e))).toList(),
      totalCount:
          readField<num>(json, 'TotalCount')?.toInt() ??
          readField<num>(json, 'totalCount')?.toInt() ??
          0,
      page:
          readField<num>(json, 'Page')?.toInt() ??
          readField<num>(json, 'page')?.toInt() ??
          1,
      pageSize:
          readField<num>(json, 'PageSize')?.toInt() ??
          readField<num>(json, 'pageSize')?.toInt() ??
          10,
      hasPrevious:
          readField<bool>(json, 'HasPrevious') ??
          readField<bool>(json, 'hasPrevious') ??
          false,
      hasNext:
          readField<bool>(json, 'HasNext') ??
          readField<bool>(json, 'hasNext') ??
          false,
    );
  }
}

File: BackendApi/Core/Models/Response/ApiResponse.cs
namespace BackendApi.Core.Models.Response;

/// <summary>
/// Standard API Response wrapper — ห่อทุก Response ให้อยู่ในรูปแบบเดียวกัน
/// </summary>
public class ApiResponse
{
    public int Status { get; set; }
    public bool Success { get; set; }
    public string? Message { get; set; }
    public string? ErrorDetail { get; set; }
    public string? Code { get; set; }
    public Dictionary<string, string[]>? Errors { get; set; }

    public static ApiResponse Ok(string? message = null) =>
        new() { Status = StatusCodes.Status200OK, Success = true, Message = message ?? "สำเร็จ" };

    public static ApiResponse Fail(string message, string? errorDetail = null, string? code = null) =>
        Fail(StatusCodes.Status400BadRequest, message, errorDetail, code);

    public static ApiResponse Fail(
        int status,
        string message,
        string? errorDetail = null,
        string? code = null,
        Dictionary<string, string[]>? errors = null) =>
        new()
        {
            Status = status,
            Success = false,
            Message = message,
            ErrorDetail = errorDetail,
            Code = code,
            Errors = errors
        };
}

/// <summary>
/// Standard API Response wrapper with payload
/// </summary>
public class ApiResponse<T> : ApiResponse
{
    public T? Value { get; set; }

    public static ApiResponse<T> Ok(T value, string? message = null) =>
        new()
        {
            Status = StatusCodes.Status200OK,
            Success = true,
            Value = value,
            Message = message ?? "สำเร็จ"
        };

    public new static ApiResponse<T> Fail(string message, string? errorDetail = null, string? code = null) =>
        Fail(StatusCodes.Status400BadRequest, message, errorDetail, code);

    public new static ApiResponse<T> Fail(
        int status,
        string message,
        string? errorDetail = null,
        string? code = null,
        Dictionary<string, string[]>? errors = null) =>
        new()
        {
            Status = status,
            Success = false,
            Message = message,
            ErrorDetail = errorDetail,
            Code = code,
            Errors = errors
        };
}

